content/chainguard/migration-guides/tutorial/index.md

```diff
@@ -0,0 +1,732 @@
+Academy guide to Python
+images](https://edu.chainguard.dev/chainguard/chainguard-images/getting-started/python/).
+
+Replace the Dockerfile with this one (this is also available on the "main" branch):
+
+```Dockerfile
+FROM cgr.dev/chainguard/python:latest-dev as dev
+
+ENV LANG=C.UTF-8
+ENV PYTHONDONTWRITEBYTECODE=1
+ENV PYTHONUNBUFFERED=1
+ENV PATH="/app/venv/bin:$PATH"
+
+WORKDIR /app
+RUN python -m venv /app/venv
+COPY requirements.txt ./
+RUN pip install --no-cache-dir -r requirements.txt
+COPY app /app
+
+EXPOSE 5000
+
+ENTRYPOINT ["python"]
+CMD ["identidock.py"]
+
+FROM cgr.dev/chainguard/python
+
+WORKDIR /app
+ENV PYTHONUNBUFFERED=1
+ENV PATH="/app/venv/bin:$PATH"
+
+COPY app/identidock.py ./
+COPY --from=dev /app/venv /app/venv
+
+EXPOSE 9090
+ENTRYPOINT [ "gunicorn", "-b", "0.0.0.0:9090", "identidock:app" ]
+```
+
+And add the file `requirements.txt` to the current directory (`identidock`) with the following contents:
+
+```
+Flask==3.0.2
+requests==2.31.0
+redis==5.0.3
+gunicorn==21.2.0
+```
+
+The first thing to notice is that we have a multistage build now. If you want the development image
+rather than the production one, you can specify it during docker build:
+
+```bash
+docker build --target dev -t identidock:dev .
+```
+
+Otherwise, you will get the production image only.
+
+There are several more environment variables defined. These prevent the creation of Python bytecode
+and buffering of output. The reasons why you want these are normally set are explained in detail in
+the blog [PYTHONDONTWRITEBYTECODE and PYTHONUNBUFFERED
+Explained](https://blog.mimixtech.com/pythondontwritebytecode-and-pythonunbuffered-explained).
+
+The installation of pip modules has moved to the `requirements.txt` file. The main thinking here is
+that we don't need to update the Dockerfile each time a dependency is updated or changed.
+
+The development server runs on port 5000, while the production server runs on port 9090. We could
+edit this so they both run on the same port, but this approach reduces the chance of accidentally
+running the development server in production. The development server is started directly from the
+entrypoint, so we are no longer dependent on an entrypoint script, simplifying our architecture.
+
+To get a minimal, clean production install, we are using a Python virtual environment
+([venv](https://docs.python.org/3/library/venv.html)) in the development image to isolate all
+dependencies, which are then copied over to the production image. Finally, the production image has
+been changed to use [gunicorn](https://gunicorn.org/) as [uwsgi has entered "maintenance
+mode"](https://github.com/unbit/uwsgi).
+
+Build the final image:
+
+```bash
+docker build -t identidock-cg .
+```
+
+And take a look at it:
+
+```bash
+docker images identidock-cg
+REPOSITORY      TAG       IMAGE ID       CREATED          SIZE
+identidock-cg   latest    30a424aa8a51   15 seconds ago   93.6MB
```

Comment:
I got 83.2 MB fwiw

Reply:
I get 91.7 now :)

Reply:
I'm going to leave this as it's entirely dependent on the size of the base image on the day.